### DISCUSSION TITLE
NoSuchMethodError: com.google.common.base.Preconditions.checkArgument

### DISCUSSION THREAD

**Discussion by tantalate** [playframework/playframework#11224] [Q&A]
Hi guys, I am trying to build a website with playframework and Java, and the current play version is 2.8.15.

I got a NoSuchMethodError runtime error (when responding to a simple request) but cannot tell where the error is located from the stack trace. I failed to find any related documents on the Internet. Any one can give a hint?
The code used to work but after I added some simple controllers and routes, it won't work and give no clue of where the error come from.

After I added a JAR file in /lib folder, I got the following runtime error (occurred on any request).
The JAR file is an SDK comes from our cooperative team, and it works well independently.  The file is named as `business.sdk-1.0.2-tool-jar-with-dependencies.jar` in case the name matters.

 Does play have special requirements of the unmanaged dependencies?

```
  Unexpected exception
      NoSuchMethodError: com.google.common.base.Preconditions.checkArgument(ZLjava/lang/String;Ljava/lang/Object;)V
  
    No source available, here is the exception stack trace:
      ->java.lang.NoSuchMethodError: com.google.common.base.Preconditions.checkArgument(ZLjava/lang/String;Ljava/lang/Object;)V
    com.google.inject.Key.ensureRetainedAtRuntime(Key.java:349)
    com.google.inject.Key.strategyFor(Key.java:325)
    com.google.inject.Key.get(Key.java:232)
    play.api.inject.guice.GuiceKey$.apply(GuiceInjectorBuilder.scala:416)
    play.api.inject.guice.GuiceableModuleConversions$$anon$4.$anonfun$configure$2(GuiceInjectorBuilder.scala:376)
    play.api.inject.guice.GuiceableModuleConversions$$anon$4.$anonfun$configure$2$adapted(GuiceInjectorBuilder.scala:374)
    scala.collection.immutable.List.foreach(List.scala:333)
    play.api.inject.guice.GuiceableModuleConversions$$anon$4.configure(GuiceInjectorBuilder.scala:374)
    com.google.inject.AbstractModule.configure(AbstractModule.java:61)
    com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:347)
    com.google.inject.spi.Elements.getElements(Elements.java:104)
    com.google.inject.util.Modules$OverrideModule.configure(Modules.java:214)
    com.google.inject.AbstractModule.configure(AbstractModule.java:61)
    com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:347)
    com.google.inject.spi.Elements.getElements(Elements.java:104)
    com.google.inject.internal.InjectorShell$Builder.build(InjectorShell.java:137)
    com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:105)
    com.google.inject.Guice.createInjector(Guice.java:87)
    com.google.inject.Guice.createInjector(Guice.java:78)
    play.api.inject.guice.GuiceBuilder.injector(GuiceInjectorBuilder.scala:200)
    play.api.inject.guice.GuiceApplicationBuilder.build(GuiceApplicationBuilder.scala:155)
    play.api.inject.guice.GuiceApplicationLoader.load(GuiceApplicationLoader.scala:21)
    play.core.server.DevServerStart$$anon$1.$anonfun$reload$2(DevServerStart.scala:193)
    play.utils.Threads$.withContextClassLoader(Threads.scala:22)
    play.core.server.DevServerStart$$anon$1.reload(DevServerStart.scala:185)
    play.core.server.DevServerStart$$anon$1.get(DevServerStart.scala:148)
    play.core.server.AkkaHttpServer.handleRequest(AkkaHttpServer.scala:302)
    play.core.server.AkkaHttpServer.$anonfun$createServerBinding$1(AkkaHttpServer.scala:224)
    akka.stream.impl.fusing.MapAsync$$anon$30.onPush(Ops.scala:1307)
    akka.stream.impl.fusing.GraphInterpreter.processPush(GraphInterpreter.scala:542)
    akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:496)
    akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:390)
    akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:650)
    akka.stream.impl.fusing.GraphInterpreterShell$AsyncInput.execute(ActorGraphInterpreter.scala:521)
    akka.stream.impl.fusing.GraphInterpreterShell.processEvent(ActorGraphInterpreter.scala:625)
    akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:800)
    akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:818)
    akka.actor.Actor.aroundReceive(Actor.scala:537)
    akka.actor.Actor.aroundReceive$(Actor.scala:535)
    akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:716)
    akka.actor.ActorCell.receiveMessage(ActorCell.scala:580)
    akka.actor.ActorCell.invoke(ActorCell.scala:548)
    akka.dispatch.Mailbox.processMailbox(Mailbox.scala:270)
    akka.dispatch.Mailbox.run(Mailbox.scala:231)
    akka.dispatch.Mailbox.exec(Mailbox.scala:243)
    java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
    java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
    java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
    java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
    java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

**Comment by mkurz** [comment]
The name of the jar your are adding indicates that it contains dependencies: `business.sdk-1.0.2-tool-jar-with-dependencies.jar`.
I am pretty sure the dependencies it includes conflict with the dependencies Play pulls in and this is causing the problem.
Does your cooperative team also provide a jar *without* the dependencies? Like`business.sdk-1.0.2-tool-jar-no-dependencies.jar`?
Also
> it works well independently

Do you mean when you run it with `java -jar ...`?

**Reply by tantalate** [reply]
I have looked into the decompiled classes of this jar, and it has a com.google.guava included, which I think might cause the problem. However, I have tried different versions of guava by `dependencyOverrides`, and the error still occurs. By the way, `sbt evicted` don't show the dependencies of the jar. 

Is this error caused by code from play or inside the jar?  This occurs even if I don't import any classes from the jar.

**Comment by mkurz** [comment]
You mean this also occurs if your `/lib` folder is empty? Did you also delete the `/target` folder in the project before you tried again?

**Reply by tantalate** [reply]
I mean the error occurs after I put the jar in /lib folder, before I use any classes from the jar. What should I do if the reason is as what you said, the jar comes with another version of guava? I have tried with specifying the version but no effect. 

I am also asking the SDK provider to give one without dependencies.

**Reply by mkurz** [reply]
Play 2.8.15 comes with guava `30.1.1-jre`:
https://github.com/playframework/playframework/blob/a898ec0f88246f1d20f5af698d7d0b6df7c1b366/project/Dependencies.scala#L56
I don't know which guava version your lib provides (Do you know?). Best thing is to reach out the SDK provider and ask them if they can provide the library with compatibility with the guava version Play is using.

**Reply by tantalate** [reply]
They are using 19.0 according to their dependency tree. Is it possible to leave their version in their JAR without interfering play framework? I think maven might be able to achieve this by excluding the package from the jar.

**Reply by tantalate** [reply]
By the way, I have already tried to specify the version to 30.1.1-jre with`dependencyOverrides`. Does that mean the same as maven do with an exclusion?

**Reply by mkurz** [reply]
> By the way, I have already tried to specify the version to 30.1.1-jre with `dependencyOverrides`.

That doesn't make sense, because Play ships with guava 30.1.1-jre anyway. So you change nothing with this config.
The problem you are facing is that the library you are using depends on a very old guava release ([from 2015](https://github.com/google/guava/releases/tag/v19.0)). The thing is not just Play, but also Guice depends on a recent guava release, so even if you exclude guava from Play you will have problems with Guice.
The only choice IMHO you have is to tell the SDK devs to upgrade to latest guava (or completely get rid of it).
You might get lucky removing the dependency from the jar file yourself.

However I think there is nothing I can do to help you here from the Play side.

**Reply by tantalate** [reply]
Are you saying the reason is guava 30.1.1-jre is not compatible with the jar file, and some piece of code in this jar file calls a method that is only implemented in this old version of guava? If that is the case, should this error only occur when this piece of code is running? As I mentioned, the error occurs before I import any classes from that jar. 
Also as a test, I started a new project, with guava 30.1.1-jre added to the dependencies list (with maven not sbt, I am new to sbt too).  In this project, I can import classes from this jar file, and it run correctly.

**Reply by mkurz** [reply]
> Are you saying the reason is guava 30.1.1-jre is not compatible with the jar file, and some piece of code in this jar file calls a method that is only implemented in this old version of guava?

I would say the guava version your sdk library ships with is not compatible with Play. Also I would guess that the guava version Play ships does not work with the library you are trying to add.

> If that is the case, should this error only occur when this piece of code is running?

That is what is happening... When Play tries to inject a component via DI it calls some guava method which is failing. The reason why it fails is because it looks like the classloader loaded the guava classes from your sdk library first (which Play is incompatible with).

> Also as a test, I started a new project, with guava 30.1.1-jre added to the dependencies list (with maven not sbt, I am new to sbt too). In this project, I can import classes from this jar file, and it run correctly.

That might be just luck. I am not 100%, but I would guess maven just puts the jar on the classpath in the "correct" order ("correct" meaning that you are lucky and it puts the compatible jar first).
